<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns='http://www.w3.org/1999/xhtml' xml:lang='en' lang='en'>
<head>
<title>bluez.git - bluez</title>
<meta name='generator' content='cgit v0.9.0.3'/>
<meta name='robots' content='index, nofollow'/>
<link rel='stylesheet' type='text/css' href='/cgit-data/cgit.css'/>
<link rel='alternate' title='Atom feed' href='http://pkgs.fedoraproject.org/cgit/bluez.git/atom/0001-Add-sixaxis-cable-pairing-plugin.patch?h=master' type='application/atom+xml'/>
</head>
<body>
<div id='cgit'><table id='header'>
<tr>
<td class='logo' rowspan='2'><a href='/cgit/'><img src='/cgit-data/cgit.png' alt='cgit logo'/></a></td>
<td class='main'><a href='/cgit/'>index</a> : <a title='bluez.git' href='/cgit/bluez.git/'>bluez.git</a></td><td class='form'><form method='get' action=''>
<select name='h' onchange='this.form.submit();'>
<option value='f10'>f10</option>
<option value='f11'>f11</option>
<option value='f12'>f12</option>
<option value='f13'>f13</option>
<option value='f14'>f14</option>
<option value='f15'>f15</option>
<option value='f16'>f16</option>
<option value='f17'>f17</option>
<option value='f18'>f18</option>
<option value='f9'>f9</option>
<option value='master' selected='selected'>master</option>
</select> <input type='submit' name='' value='switch'/></form></td></tr>
<tr><td class='sub'>bluez</td><td class='sub right'>Jesse Keating</td></tr></table>
<table class='tabs'><tr><td>
<a href='/cgit/bluez.git/'>summary</a><a href='/cgit/bluez.git/refs/'>refs</a><a href='/cgit/bluez.git/log/0001-Add-sixaxis-cable-pairing-plugin.patch'>log</a><a class='active' href='/cgit/bluez.git/tree/0001-Add-sixaxis-cable-pairing-plugin.patch'>tree</a><a href='/cgit/bluez.git/commit/0001-Add-sixaxis-cable-pairing-plugin.patch'>commit</a><a href='/cgit/bluez.git/diff/0001-Add-sixaxis-cable-pairing-plugin.patch'>diff</a><a href='/cgit/bluez.git/stats/0001-Add-sixaxis-cable-pairing-plugin.patch'>stats</a></td><td class='form'><form class='right' method='get' action='/cgit/bluez.git/log/0001-Add-sixaxis-cable-pairing-plugin.patch'>
<select name='qt'>
<option value='grep'>log msg</option>
<option value='author'>author</option>
<option value='committer'>committer</option>
<option value='range'>range</option>
</select>
<input class='txt' type='text' size='10' name='q' value=''/>
<input type='submit' value='search'/>
</form>
</td></tr></table>
<div class='path'>path: <a href='/cgit/bluez.git/tree/'>root</a>/<a href='/cgit/bluez.git/tree/0001-Add-sixaxis-cable-pairing-plugin.patch'>0001-Add-sixaxis-cable-pairing-plugin.patch</a></div><div class='content'>blob: b20c9c8f4a1b8462a6dabadec2efb78b8bc02c56 (<a href='/cgit/bluez.git/plain/0001-Add-sixaxis-cable-pairing-plugin.patch'>plain</a>)
<table summary='blob content' class='blob'>
<tr><td class='linenumbers'><pre><a class='no' id='n1' name='n1' href='#n1'>1</a>
<a class='no' id='n2' name='n2' href='#n2'>2</a>
<a class='no' id='n3' name='n3' href='#n3'>3</a>
<a class='no' id='n4' name='n4' href='#n4'>4</a>
<a class='no' id='n5' name='n5' href='#n5'>5</a>
<a class='no' id='n6' name='n6' href='#n6'>6</a>
<a class='no' id='n7' name='n7' href='#n7'>7</a>
<a class='no' id='n8' name='n8' href='#n8'>8</a>
<a class='no' id='n9' name='n9' href='#n9'>9</a>
<a class='no' id='n10' name='n10' href='#n10'>10</a>
<a class='no' id='n11' name='n11' href='#n11'>11</a>
<a class='no' id='n12' name='n12' href='#n12'>12</a>
<a class='no' id='n13' name='n13' href='#n13'>13</a>
<a class='no' id='n14' name='n14' href='#n14'>14</a>
<a class='no' id='n15' name='n15' href='#n15'>15</a>
<a class='no' id='n16' name='n16' href='#n16'>16</a>
<a class='no' id='n17' name='n17' href='#n17'>17</a>
<a class='no' id='n18' name='n18' href='#n18'>18</a>
<a class='no' id='n19' name='n19' href='#n19'>19</a>
<a class='no' id='n20' name='n20' href='#n20'>20</a>
<a class='no' id='n21' name='n21' href='#n21'>21</a>
<a class='no' id='n22' name='n22' href='#n22'>22</a>
<a class='no' id='n23' name='n23' href='#n23'>23</a>
<a class='no' id='n24' name='n24' href='#n24'>24</a>
<a class='no' id='n25' name='n25' href='#n25'>25</a>
<a class='no' id='n26' name='n26' href='#n26'>26</a>
<a class='no' id='n27' name='n27' href='#n27'>27</a>
<a class='no' id='n28' name='n28' href='#n28'>28</a>
<a class='no' id='n29' name='n29' href='#n29'>29</a>
<a class='no' id='n30' name='n30' href='#n30'>30</a>
<a class='no' id='n31' name='n31' href='#n31'>31</a>
<a class='no' id='n32' name='n32' href='#n32'>32</a>
<a class='no' id='n33' name='n33' href='#n33'>33</a>
<a class='no' id='n34' name='n34' href='#n34'>34</a>
<a class='no' id='n35' name='n35' href='#n35'>35</a>
<a class='no' id='n36' name='n36' href='#n36'>36</a>
<a class='no' id='n37' name='n37' href='#n37'>37</a>
<a class='no' id='n38' name='n38' href='#n38'>38</a>
<a class='no' id='n39' name='n39' href='#n39'>39</a>
<a class='no' id='n40' name='n40' href='#n40'>40</a>
<a class='no' id='n41' name='n41' href='#n41'>41</a>
<a class='no' id='n42' name='n42' href='#n42'>42</a>
<a class='no' id='n43' name='n43' href='#n43'>43</a>
<a class='no' id='n44' name='n44' href='#n44'>44</a>
<a class='no' id='n45' name='n45' href='#n45'>45</a>
<a class='no' id='n46' name='n46' href='#n46'>46</a>
<a class='no' id='n47' name='n47' href='#n47'>47</a>
<a class='no' id='n48' name='n48' href='#n48'>48</a>
<a class='no' id='n49' name='n49' href='#n49'>49</a>
<a class='no' id='n50' name='n50' href='#n50'>50</a>
<a class='no' id='n51' name='n51' href='#n51'>51</a>
<a class='no' id='n52' name='n52' href='#n52'>52</a>
<a class='no' id='n53' name='n53' href='#n53'>53</a>
<a class='no' id='n54' name='n54' href='#n54'>54</a>
<a class='no' id='n55' name='n55' href='#n55'>55</a>
<a class='no' id='n56' name='n56' href='#n56'>56</a>
<a class='no' id='n57' name='n57' href='#n57'>57</a>
<a class='no' id='n58' name='n58' href='#n58'>58</a>
<a class='no' id='n59' name='n59' href='#n59'>59</a>
<a class='no' id='n60' name='n60' href='#n60'>60</a>
<a class='no' id='n61' name='n61' href='#n61'>61</a>
<a class='no' id='n62' name='n62' href='#n62'>62</a>
<a class='no' id='n63' name='n63' href='#n63'>63</a>
<a class='no' id='n64' name='n64' href='#n64'>64</a>
<a class='no' id='n65' name='n65' href='#n65'>65</a>
<a class='no' id='n66' name='n66' href='#n66'>66</a>
<a class='no' id='n67' name='n67' href='#n67'>67</a>
<a class='no' id='n68' name='n68' href='#n68'>68</a>
<a class='no' id='n69' name='n69' href='#n69'>69</a>
<a class='no' id='n70' name='n70' href='#n70'>70</a>
<a class='no' id='n71' name='n71' href='#n71'>71</a>
<a class='no' id='n72' name='n72' href='#n72'>72</a>
<a class='no' id='n73' name='n73' href='#n73'>73</a>
<a class='no' id='n74' name='n74' href='#n74'>74</a>
<a class='no' id='n75' name='n75' href='#n75'>75</a>
<a class='no' id='n76' name='n76' href='#n76'>76</a>
<a class='no' id='n77' name='n77' href='#n77'>77</a>
<a class='no' id='n78' name='n78' href='#n78'>78</a>
<a class='no' id='n79' name='n79' href='#n79'>79</a>
<a class='no' id='n80' name='n80' href='#n80'>80</a>
<a class='no' id='n81' name='n81' href='#n81'>81</a>
<a class='no' id='n82' name='n82' href='#n82'>82</a>
<a class='no' id='n83' name='n83' href='#n83'>83</a>
<a class='no' id='n84' name='n84' href='#n84'>84</a>
<a class='no' id='n85' name='n85' href='#n85'>85</a>
<a class='no' id='n86' name='n86' href='#n86'>86</a>
<a class='no' id='n87' name='n87' href='#n87'>87</a>
<a class='no' id='n88' name='n88' href='#n88'>88</a>
<a class='no' id='n89' name='n89' href='#n89'>89</a>
<a class='no' id='n90' name='n90' href='#n90'>90</a>
<a class='no' id='n91' name='n91' href='#n91'>91</a>
<a class='no' id='n92' name='n92' href='#n92'>92</a>
<a class='no' id='n93' name='n93' href='#n93'>93</a>
<a class='no' id='n94' name='n94' href='#n94'>94</a>
<a class='no' id='n95' name='n95' href='#n95'>95</a>
<a class='no' id='n96' name='n96' href='#n96'>96</a>
<a class='no' id='n97' name='n97' href='#n97'>97</a>
<a class='no' id='n98' name='n98' href='#n98'>98</a>
<a class='no' id='n99' name='n99' href='#n99'>99</a>
<a class='no' id='n100' name='n100' href='#n100'>100</a>
<a class='no' id='n101' name='n101' href='#n101'>101</a>
<a class='no' id='n102' name='n102' href='#n102'>102</a>
<a class='no' id='n103' name='n103' href='#n103'>103</a>
<a class='no' id='n104' name='n104' href='#n104'>104</a>
<a class='no' id='n105' name='n105' href='#n105'>105</a>
<a class='no' id='n106' name='n106' href='#n106'>106</a>
<a class='no' id='n107' name='n107' href='#n107'>107</a>
<a class='no' id='n108' name='n108' href='#n108'>108</a>
<a class='no' id='n109' name='n109' href='#n109'>109</a>
<a class='no' id='n110' name='n110' href='#n110'>110</a>
<a class='no' id='n111' name='n111' href='#n111'>111</a>
<a class='no' id='n112' name='n112' href='#n112'>112</a>
<a class='no' id='n113' name='n113' href='#n113'>113</a>
<a class='no' id='n114' name='n114' href='#n114'>114</a>
<a class='no' id='n115' name='n115' href='#n115'>115</a>
<a class='no' id='n116' name='n116' href='#n116'>116</a>
<a class='no' id='n117' name='n117' href='#n117'>117</a>
<a class='no' id='n118' name='n118' href='#n118'>118</a>
<a class='no' id='n119' name='n119' href='#n119'>119</a>
<a class='no' id='n120' name='n120' href='#n120'>120</a>
<a class='no' id='n121' name='n121' href='#n121'>121</a>
<a class='no' id='n122' name='n122' href='#n122'>122</a>
<a class='no' id='n123' name='n123' href='#n123'>123</a>
<a class='no' id='n124' name='n124' href='#n124'>124</a>
<a class='no' id='n125' name='n125' href='#n125'>125</a>
<a class='no' id='n126' name='n126' href='#n126'>126</a>
<a class='no' id='n127' name='n127' href='#n127'>127</a>
<a class='no' id='n128' name='n128' href='#n128'>128</a>
<a class='no' id='n129' name='n129' href='#n129'>129</a>
<a class='no' id='n130' name='n130' href='#n130'>130</a>
<a class='no' id='n131' name='n131' href='#n131'>131</a>
<a class='no' id='n132' name='n132' href='#n132'>132</a>
<a class='no' id='n133' name='n133' href='#n133'>133</a>
<a class='no' id='n134' name='n134' href='#n134'>134</a>
<a class='no' id='n135' name='n135' href='#n135'>135</a>
<a class='no' id='n136' name='n136' href='#n136'>136</a>
<a class='no' id='n137' name='n137' href='#n137'>137</a>
<a class='no' id='n138' name='n138' href='#n138'>138</a>
<a class='no' id='n139' name='n139' href='#n139'>139</a>
<a class='no' id='n140' name='n140' href='#n140'>140</a>
<a class='no' id='n141' name='n141' href='#n141'>141</a>
<a class='no' id='n142' name='n142' href='#n142'>142</a>
<a class='no' id='n143' name='n143' href='#n143'>143</a>
<a class='no' id='n144' name='n144' href='#n144'>144</a>
<a class='no' id='n145' name='n145' href='#n145'>145</a>
<a class='no' id='n146' name='n146' href='#n146'>146</a>
<a class='no' id='n147' name='n147' href='#n147'>147</a>
<a class='no' id='n148' name='n148' href='#n148'>148</a>
<a class='no' id='n149' name='n149' href='#n149'>149</a>
<a class='no' id='n150' name='n150' href='#n150'>150</a>
<a class='no' id='n151' name='n151' href='#n151'>151</a>
<a class='no' id='n152' name='n152' href='#n152'>152</a>
<a class='no' id='n153' name='n153' href='#n153'>153</a>
<a class='no' id='n154' name='n154' href='#n154'>154</a>
<a class='no' id='n155' name='n155' href='#n155'>155</a>
<a class='no' id='n156' name='n156' href='#n156'>156</a>
<a class='no' id='n157' name='n157' href='#n157'>157</a>
<a class='no' id='n158' name='n158' href='#n158'>158</a>
<a class='no' id='n159' name='n159' href='#n159'>159</a>
<a class='no' id='n160' name='n160' href='#n160'>160</a>
<a class='no' id='n161' name='n161' href='#n161'>161</a>
<a class='no' id='n162' name='n162' href='#n162'>162</a>
<a class='no' id='n163' name='n163' href='#n163'>163</a>
<a class='no' id='n164' name='n164' href='#n164'>164</a>
<a class='no' id='n165' name='n165' href='#n165'>165</a>
<a class='no' id='n166' name='n166' href='#n166'>166</a>
<a class='no' id='n167' name='n167' href='#n167'>167</a>
<a class='no' id='n168' name='n168' href='#n168'>168</a>
<a class='no' id='n169' name='n169' href='#n169'>169</a>
<a class='no' id='n170' name='n170' href='#n170'>170</a>
<a class='no' id='n171' name='n171' href='#n171'>171</a>
<a class='no' id='n172' name='n172' href='#n172'>172</a>
<a class='no' id='n173' name='n173' href='#n173'>173</a>
<a class='no' id='n174' name='n174' href='#n174'>174</a>
<a class='no' id='n175' name='n175' href='#n175'>175</a>
<a class='no' id='n176' name='n176' href='#n176'>176</a>
<a class='no' id='n177' name='n177' href='#n177'>177</a>
<a class='no' id='n178' name='n178' href='#n178'>178</a>
<a class='no' id='n179' name='n179' href='#n179'>179</a>
<a class='no' id='n180' name='n180' href='#n180'>180</a>
<a class='no' id='n181' name='n181' href='#n181'>181</a>
<a class='no' id='n182' name='n182' href='#n182'>182</a>
<a class='no' id='n183' name='n183' href='#n183'>183</a>
<a class='no' id='n184' name='n184' href='#n184'>184</a>
<a class='no' id='n185' name='n185' href='#n185'>185</a>
<a class='no' id='n186' name='n186' href='#n186'>186</a>
<a class='no' id='n187' name='n187' href='#n187'>187</a>
<a class='no' id='n188' name='n188' href='#n188'>188</a>
<a class='no' id='n189' name='n189' href='#n189'>189</a>
<a class='no' id='n190' name='n190' href='#n190'>190</a>
<a class='no' id='n191' name='n191' href='#n191'>191</a>
<a class='no' id='n192' name='n192' href='#n192'>192</a>
<a class='no' id='n193' name='n193' href='#n193'>193</a>
<a class='no' id='n194' name='n194' href='#n194'>194</a>
<a class='no' id='n195' name='n195' href='#n195'>195</a>
<a class='no' id='n196' name='n196' href='#n196'>196</a>
<a class='no' id='n197' name='n197' href='#n197'>197</a>
<a class='no' id='n198' name='n198' href='#n198'>198</a>
<a class='no' id='n199' name='n199' href='#n199'>199</a>
<a class='no' id='n200' name='n200' href='#n200'>200</a>
<a class='no' id='n201' name='n201' href='#n201'>201</a>
<a class='no' id='n202' name='n202' href='#n202'>202</a>
<a class='no' id='n203' name='n203' href='#n203'>203</a>
<a class='no' id='n204' name='n204' href='#n204'>204</a>
<a class='no' id='n205' name='n205' href='#n205'>205</a>
<a class='no' id='n206' name='n206' href='#n206'>206</a>
<a class='no' id='n207' name='n207' href='#n207'>207</a>
<a class='no' id='n208' name='n208' href='#n208'>208</a>
<a class='no' id='n209' name='n209' href='#n209'>209</a>
<a class='no' id='n210' name='n210' href='#n210'>210</a>
<a class='no' id='n211' name='n211' href='#n211'>211</a>
<a class='no' id='n212' name='n212' href='#n212'>212</a>
<a class='no' id='n213' name='n213' href='#n213'>213</a>
<a class='no' id='n214' name='n214' href='#n214'>214</a>
<a class='no' id='n215' name='n215' href='#n215'>215</a>
<a class='no' id='n216' name='n216' href='#n216'>216</a>
<a class='no' id='n217' name='n217' href='#n217'>217</a>
<a class='no' id='n218' name='n218' href='#n218'>218</a>
<a class='no' id='n219' name='n219' href='#n219'>219</a>
<a class='no' id='n220' name='n220' href='#n220'>220</a>
<a class='no' id='n221' name='n221' href='#n221'>221</a>
<a class='no' id='n222' name='n222' href='#n222'>222</a>
<a class='no' id='n223' name='n223' href='#n223'>223</a>
<a class='no' id='n224' name='n224' href='#n224'>224</a>
<a class='no' id='n225' name='n225' href='#n225'>225</a>
<a class='no' id='n226' name='n226' href='#n226'>226</a>
<a class='no' id='n227' name='n227' href='#n227'>227</a>
<a class='no' id='n228' name='n228' href='#n228'>228</a>
<a class='no' id='n229' name='n229' href='#n229'>229</a>
<a class='no' id='n230' name='n230' href='#n230'>230</a>
<a class='no' id='n231' name='n231' href='#n231'>231</a>
<a class='no' id='n232' name='n232' href='#n232'>232</a>
<a class='no' id='n233' name='n233' href='#n233'>233</a>
<a class='no' id='n234' name='n234' href='#n234'>234</a>
<a class='no' id='n235' name='n235' href='#n235'>235</a>
<a class='no' id='n236' name='n236' href='#n236'>236</a>
<a class='no' id='n237' name='n237' href='#n237'>237</a>
<a class='no' id='n238' name='n238' href='#n238'>238</a>
<a class='no' id='n239' name='n239' href='#n239'>239</a>
<a class='no' id='n240' name='n240' href='#n240'>240</a>
<a class='no' id='n241' name='n241' href='#n241'>241</a>
<a class='no' id='n242' name='n242' href='#n242'>242</a>
<a class='no' id='n243' name='n243' href='#n243'>243</a>
<a class='no' id='n244' name='n244' href='#n244'>244</a>
<a class='no' id='n245' name='n245' href='#n245'>245</a>
<a class='no' id='n246' name='n246' href='#n246'>246</a>
<a class='no' id='n247' name='n247' href='#n247'>247</a>
<a class='no' id='n248' name='n248' href='#n248'>248</a>
<a class='no' id='n249' name='n249' href='#n249'>249</a>
<a class='no' id='n250' name='n250' href='#n250'>250</a>
<a class='no' id='n251' name='n251' href='#n251'>251</a>
<a class='no' id='n252' name='n252' href='#n252'>252</a>
<a class='no' id='n253' name='n253' href='#n253'>253</a>
<a class='no' id='n254' name='n254' href='#n254'>254</a>
<a class='no' id='n255' name='n255' href='#n255'>255</a>
<a class='no' id='n256' name='n256' href='#n256'>256</a>
<a class='no' id='n257' name='n257' href='#n257'>257</a>
<a class='no' id='n258' name='n258' href='#n258'>258</a>
<a class='no' id='n259' name='n259' href='#n259'>259</a>
<a class='no' id='n260' name='n260' href='#n260'>260</a>
<a class='no' id='n261' name='n261' href='#n261'>261</a>
<a class='no' id='n262' name='n262' href='#n262'>262</a>
<a class='no' id='n263' name='n263' href='#n263'>263</a>
<a class='no' id='n264' name='n264' href='#n264'>264</a>
<a class='no' id='n265' name='n265' href='#n265'>265</a>
<a class='no' id='n266' name='n266' href='#n266'>266</a>
<a class='no' id='n267' name='n267' href='#n267'>267</a>
<a class='no' id='n268' name='n268' href='#n268'>268</a>
<a class='no' id='n269' name='n269' href='#n269'>269</a>
<a class='no' id='n270' name='n270' href='#n270'>270</a>
<a class='no' id='n271' name='n271' href='#n271'>271</a>
<a class='no' id='n272' name='n272' href='#n272'>272</a>
<a class='no' id='n273' name='n273' href='#n273'>273</a>
<a class='no' id='n274' name='n274' href='#n274'>274</a>
<a class='no' id='n275' name='n275' href='#n275'>275</a>
<a class='no' id='n276' name='n276' href='#n276'>276</a>
<a class='no' id='n277' name='n277' href='#n277'>277</a>
<a class='no' id='n278' name='n278' href='#n278'>278</a>
<a class='no' id='n279' name='n279' href='#n279'>279</a>
<a class='no' id='n280' name='n280' href='#n280'>280</a>
<a class='no' id='n281' name='n281' href='#n281'>281</a>
<a class='no' id='n282' name='n282' href='#n282'>282</a>
<a class='no' id='n283' name='n283' href='#n283'>283</a>
<a class='no' id='n284' name='n284' href='#n284'>284</a>
<a class='no' id='n285' name='n285' href='#n285'>285</a>
<a class='no' id='n286' name='n286' href='#n286'>286</a>
<a class='no' id='n287' name='n287' href='#n287'>287</a>
<a class='no' id='n288' name='n288' href='#n288'>288</a>
<a class='no' id='n289' name='n289' href='#n289'>289</a>
<a class='no' id='n290' name='n290' href='#n290'>290</a>
<a class='no' id='n291' name='n291' href='#n291'>291</a>
<a class='no' id='n292' name='n292' href='#n292'>292</a>
<a class='no' id='n293' name='n293' href='#n293'>293</a>
<a class='no' id='n294' name='n294' href='#n294'>294</a>
<a class='no' id='n295' name='n295' href='#n295'>295</a>
<a class='no' id='n296' name='n296' href='#n296'>296</a>
<a class='no' id='n297' name='n297' href='#n297'>297</a>
<a class='no' id='n298' name='n298' href='#n298'>298</a>
<a class='no' id='n299' name='n299' href='#n299'>299</a>
<a class='no' id='n300' name='n300' href='#n300'>300</a>
<a class='no' id='n301' name='n301' href='#n301'>301</a>
<a class='no' id='n302' name='n302' href='#n302'>302</a>
<a class='no' id='n303' name='n303' href='#n303'>303</a>
<a class='no' id='n304' name='n304' href='#n304'>304</a>
<a class='no' id='n305' name='n305' href='#n305'>305</a>
<a class='no' id='n306' name='n306' href='#n306'>306</a>
<a class='no' id='n307' name='n307' href='#n307'>307</a>
<a class='no' id='n308' name='n308' href='#n308'>308</a>
<a class='no' id='n309' name='n309' href='#n309'>309</a>
<a class='no' id='n310' name='n310' href='#n310'>310</a>
<a class='no' id='n311' name='n311' href='#n311'>311</a>
<a class='no' id='n312' name='n312' href='#n312'>312</a>
<a class='no' id='n313' name='n313' href='#n313'>313</a>
<a class='no' id='n314' name='n314' href='#n314'>314</a>
<a class='no' id='n315' name='n315' href='#n315'>315</a>
<a class='no' id='n316' name='n316' href='#n316'>316</a>
<a class='no' id='n317' name='n317' href='#n317'>317</a>
<a class='no' id='n318' name='n318' href='#n318'>318</a>
<a class='no' id='n319' name='n319' href='#n319'>319</a>
<a class='no' id='n320' name='n320' href='#n320'>320</a>
<a class='no' id='n321' name='n321' href='#n321'>321</a>
<a class='no' id='n322' name='n322' href='#n322'>322</a>
<a class='no' id='n323' name='n323' href='#n323'>323</a>
<a class='no' id='n324' name='n324' href='#n324'>324</a>
<a class='no' id='n325' name='n325' href='#n325'>325</a>
<a class='no' id='n326' name='n326' href='#n326'>326</a>
<a class='no' id='n327' name='n327' href='#n327'>327</a>
<a class='no' id='n328' name='n328' href='#n328'>328</a>
<a class='no' id='n329' name='n329' href='#n329'>329</a>
<a class='no' id='n330' name='n330' href='#n330'>330</a>
<a class='no' id='n331' name='n331' href='#n331'>331</a>
<a class='no' id='n332' name='n332' href='#n332'>332</a>
<a class='no' id='n333' name='n333' href='#n333'>333</a>
<a class='no' id='n334' name='n334' href='#n334'>334</a>
<a class='no' id='n335' name='n335' href='#n335'>335</a>
<a class='no' id='n336' name='n336' href='#n336'>336</a>
<a class='no' id='n337' name='n337' href='#n337'>337</a>
<a class='no' id='n338' name='n338' href='#n338'>338</a>
<a class='no' id='n339' name='n339' href='#n339'>339</a>
<a class='no' id='n340' name='n340' href='#n340'>340</a>
<a class='no' id='n341' name='n341' href='#n341'>341</a>
<a class='no' id='n342' name='n342' href='#n342'>342</a>
<a class='no' id='n343' name='n343' href='#n343'>343</a>
<a class='no' id='n344' name='n344' href='#n344'>344</a>
<a class='no' id='n345' name='n345' href='#n345'>345</a>
<a class='no' id='n346' name='n346' href='#n346'>346</a>
<a class='no' id='n347' name='n347' href='#n347'>347</a>
<a class='no' id='n348' name='n348' href='#n348'>348</a>
<a class='no' id='n349' name='n349' href='#n349'>349</a>
<a class='no' id='n350' name='n350' href='#n350'>350</a>
<a class='no' id='n351' name='n351' href='#n351'>351</a>
<a class='no' id='n352' name='n352' href='#n352'>352</a>
<a class='no' id='n353' name='n353' href='#n353'>353</a>
<a class='no' id='n354' name='n354' href='#n354'>354</a>
<a class='no' id='n355' name='n355' href='#n355'>355</a>
<a class='no' id='n356' name='n356' href='#n356'>356</a>
<a class='no' id='n357' name='n357' href='#n357'>357</a>
<a class='no' id='n358' name='n358' href='#n358'>358</a>
<a class='no' id='n359' name='n359' href='#n359'>359</a>
<a class='no' id='n360' name='n360' href='#n360'>360</a>
<a class='no' id='n361' name='n361' href='#n361'>361</a>
<a class='no' id='n362' name='n362' href='#n362'>362</a>
<a class='no' id='n363' name='n363' href='#n363'>363</a>
<a class='no' id='n364' name='n364' href='#n364'>364</a>
<a class='no' id='n365' name='n365' href='#n365'>365</a>
<a class='no' id='n366' name='n366' href='#n366'>366</a>
<a class='no' id='n367' name='n367' href='#n367'>367</a>
<a class='no' id='n368' name='n368' href='#n368'>368</a>
<a class='no' id='n369' name='n369' href='#n369'>369</a>
<a class='no' id='n370' name='n370' href='#n370'>370</a>
<a class='no' id='n371' name='n371' href='#n371'>371</a>
<a class='no' id='n372' name='n372' href='#n372'>372</a>
<a class='no' id='n373' name='n373' href='#n373'>373</a>
<a class='no' id='n374' name='n374' href='#n374'>374</a>
<a class='no' id='n375' name='n375' href='#n375'>375</a>
<a class='no' id='n376' name='n376' href='#n376'>376</a>
<a class='no' id='n377' name='n377' href='#n377'>377</a>
<a class='no' id='n378' name='n378' href='#n378'>378</a>
<a class='no' id='n379' name='n379' href='#n379'>379</a>
<a class='no' id='n380' name='n380' href='#n380'>380</a>
<a class='no' id='n381' name='n381' href='#n381'>381</a>
<a class='no' id='n382' name='n382' href='#n382'>382</a>
<a class='no' id='n383' name='n383' href='#n383'>383</a>
<a class='no' id='n384' name='n384' href='#n384'>384</a>
<a class='no' id='n385' name='n385' href='#n385'>385</a>
<a class='no' id='n386' name='n386' href='#n386'>386</a>
<a class='no' id='n387' name='n387' href='#n387'>387</a>
<a class='no' id='n388' name='n388' href='#n388'>388</a>
<a class='no' id='n389' name='n389' href='#n389'>389</a>
<a class='no' id='n390' name='n390' href='#n390'>390</a>
<a class='no' id='n391' name='n391' href='#n391'>391</a>
<a class='no' id='n392' name='n392' href='#n392'>392</a>
<a class='no' id='n393' name='n393' href='#n393'>393</a>
<a class='no' id='n394' name='n394' href='#n394'>394</a>
<a class='no' id='n395' name='n395' href='#n395'>395</a>
<a class='no' id='n396' name='n396' href='#n396'>396</a>
<a class='no' id='n397' name='n397' href='#n397'>397</a>
<a class='no' id='n398' name='n398' href='#n398'>398</a>
<a class='no' id='n399' name='n399' href='#n399'>399</a>
<a class='no' id='n400' name='n400' href='#n400'>400</a>
<a class='no' id='n401' name='n401' href='#n401'>401</a>
<a class='no' id='n402' name='n402' href='#n402'>402</a>
<a class='no' id='n403' name='n403' href='#n403'>403</a>
<a class='no' id='n404' name='n404' href='#n404'>404</a>
<a class='no' id='n405' name='n405' href='#n405'>405</a>
<a class='no' id='n406' name='n406' href='#n406'>406</a>
<a class='no' id='n407' name='n407' href='#n407'>407</a>
<a class='no' id='n408' name='n408' href='#n408'>408</a>
<a class='no' id='n409' name='n409' href='#n409'>409</a>
<a class='no' id='n410' name='n410' href='#n410'>410</a>
<a class='no' id='n411' name='n411' href='#n411'>411</a>
<a class='no' id='n412' name='n412' href='#n412'>412</a>
<a class='no' id='n413' name='n413' href='#n413'>413</a>
<a class='no' id='n414' name='n414' href='#n414'>414</a>
<a class='no' id='n415' name='n415' href='#n415'>415</a>
<a class='no' id='n416' name='n416' href='#n416'>416</a>
<a class='no' id='n417' name='n417' href='#n417'>417</a>
<a class='no' id='n418' name='n418' href='#n418'>418</a>
<a class='no' id='n419' name='n419' href='#n419'>419</a>
<a class='no' id='n420' name='n420' href='#n420'>420</a>
<a class='no' id='n421' name='n421' href='#n421'>421</a>
<a class='no' id='n422' name='n422' href='#n422'>422</a>
<a class='no' id='n423' name='n423' href='#n423'>423</a>
<a class='no' id='n424' name='n424' href='#n424'>424</a>
<a class='no' id='n425' name='n425' href='#n425'>425</a>
<a class='no' id='n426' name='n426' href='#n426'>426</a>
<a class='no' id='n427' name='n427' href='#n427'>427</a>
<a class='no' id='n428' name='n428' href='#n428'>428</a>
<a class='no' id='n429' name='n429' href='#n429'>429</a>
<a class='no' id='n430' name='n430' href='#n430'>430</a>
<a class='no' id='n431' name='n431' href='#n431'>431</a>
<a class='no' id='n432' name='n432' href='#n432'>432</a>
<a class='no' id='n433' name='n433' href='#n433'>433</a>
<a class='no' id='n434' name='n434' href='#n434'>434</a>
<a class='no' id='n435' name='n435' href='#n435'>435</a>
<a class='no' id='n436' name='n436' href='#n436'>436</a>
<a class='no' id='n437' name='n437' href='#n437'>437</a>
<a class='no' id='n438' name='n438' href='#n438'>438</a>
<a class='no' id='n439' name='n439' href='#n439'>439</a>
<a class='no' id='n440' name='n440' href='#n440'>440</a>
<a class='no' id='n441' name='n441' href='#n441'>441</a>
<a class='no' id='n442' name='n442' href='#n442'>442</a>
<a class='no' id='n443' name='n443' href='#n443'>443</a>
<a class='no' id='n444' name='n444' href='#n444'>444</a>
<a class='no' id='n445' name='n445' href='#n445'>445</a>
<a class='no' id='n446' name='n446' href='#n446'>446</a>
<a class='no' id='n447' name='n447' href='#n447'>447</a>
<a class='no' id='n448' name='n448' href='#n448'>448</a>
<a class='no' id='n449' name='n449' href='#n449'>449</a>
<a class='no' id='n450' name='n450' href='#n450'>450</a>
<a class='no' id='n451' name='n451' href='#n451'>451</a>
<a class='no' id='n452' name='n452' href='#n452'>452</a>
<a class='no' id='n453' name='n453' href='#n453'>453</a>
<a class='no' id='n454' name='n454' href='#n454'>454</a>
<a class='no' id='n455' name='n455' href='#n455'>455</a>
<a class='no' id='n456' name='n456' href='#n456'>456</a>
<a class='no' id='n457' name='n457' href='#n457'>457</a>
<a class='no' id='n458' name='n458' href='#n458'>458</a>
<a class='no' id='n459' name='n459' href='#n459'>459</a>
<a class='no' id='n460' name='n460' href='#n460'>460</a>
<a class='no' id='n461' name='n461' href='#n461'>461</a>
<a class='no' id='n462' name='n462' href='#n462'>462</a>
<a class='no' id='n463' name='n463' href='#n463'>463</a>
<a class='no' id='n464' name='n464' href='#n464'>464</a>
<a class='no' id='n465' name='n465' href='#n465'>465</a>
<a class='no' id='n466' name='n466' href='#n466'>466</a>
<a class='no' id='n467' name='n467' href='#n467'>467</a>
<a class='no' id='n468' name='n468' href='#n468'>468</a>
<a class='no' id='n469' name='n469' href='#n469'>469</a>
<a class='no' id='n470' name='n470' href='#n470'>470</a>
<a class='no' id='n471' name='n471' href='#n471'>471</a>
<a class='no' id='n472' name='n472' href='#n472'>472</a>
<a class='no' id='n473' name='n473' href='#n473'>473</a>
<a class='no' id='n474' name='n474' href='#n474'>474</a>
<a class='no' id='n475' name='n475' href='#n475'>475</a>
<a class='no' id='n476' name='n476' href='#n476'>476</a>
<a class='no' id='n477' name='n477' href='#n477'>477</a>
<a class='no' id='n478' name='n478' href='#n478'>478</a>
<a class='no' id='n479' name='n479' href='#n479'>479</a>
<a class='no' id='n480' name='n480' href='#n480'>480</a>
<a class='no' id='n481' name='n481' href='#n481'>481</a>
<a class='no' id='n482' name='n482' href='#n482'>482</a>
<a class='no' id='n483' name='n483' href='#n483'>483</a>
<a class='no' id='n484' name='n484' href='#n484'>484</a>
<a class='no' id='n485' name='n485' href='#n485'>485</a>
<a class='no' id='n486' name='n486' href='#n486'>486</a>
<a class='no' id='n487' name='n487' href='#n487'>487</a>
<a class='no' id='n488' name='n488' href='#n488'>488</a>
<a class='no' id='n489' name='n489' href='#n489'>489</a>
<a class='no' id='n490' name='n490' href='#n490'>490</a>
<a class='no' id='n491' name='n491' href='#n491'>491</a>
<a class='no' id='n492' name='n492' href='#n492'>492</a>
<a class='no' id='n493' name='n493' href='#n493'>493</a>
<a class='no' id='n494' name='n494' href='#n494'>494</a>
<a class='no' id='n495' name='n495' href='#n495'>495</a>
<a class='no' id='n496' name='n496' href='#n496'>496</a>
<a class='no' id='n497' name='n497' href='#n497'>497</a>
<a class='no' id='n498' name='n498' href='#n498'>498</a>
<a class='no' id='n499' name='n499' href='#n499'>499</a>
<a class='no' id='n500' name='n500' href='#n500'>500</a>
<a class='no' id='n501' name='n501' href='#n501'>501</a>
<a class='no' id='n502' name='n502' href='#n502'>502</a>
<a class='no' id='n503' name='n503' href='#n503'>503</a>
<a class='no' id='n504' name='n504' href='#n504'>504</a>
<a class='no' id='n505' name='n505' href='#n505'>505</a>
<a class='no' id='n506' name='n506' href='#n506'>506</a>
<a class='no' id='n507' name='n507' href='#n507'>507</a>
<a class='no' id='n508' name='n508' href='#n508'>508</a>
<a class='no' id='n509' name='n509' href='#n509'>509</a>
<a class='no' id='n510' name='n510' href='#n510'>510</a>
<a class='no' id='n511' name='n511' href='#n511'>511</a>
<a class='no' id='n512' name='n512' href='#n512'>512</a>
<a class='no' id='n513' name='n513' href='#n513'>513</a>
<a class='no' id='n514' name='n514' href='#n514'>514</a>
<a class='no' id='n515' name='n515' href='#n515'>515</a>
<a class='no' id='n516' name='n516' href='#n516'>516</a>
<a class='no' id='n517' name='n517' href='#n517'>517</a>
<a class='no' id='n518' name='n518' href='#n518'>518</a>
<a class='no' id='n519' name='n519' href='#n519'>519</a>
<a class='no' id='n520' name='n520' href='#n520'>520</a>
<a class='no' id='n521' name='n521' href='#n521'>521</a>
<a class='no' id='n522' name='n522' href='#n522'>522</a>
<a class='no' id='n523' name='n523' href='#n523'>523</a>
<a class='no' id='n524' name='n524' href='#n524'>524</a>
<a class='no' id='n525' name='n525' href='#n525'>525</a>
<a class='no' id='n526' name='n526' href='#n526'>526</a>
<a class='no' id='n527' name='n527' href='#n527'>527</a>
<a class='no' id='n528' name='n528' href='#n528'>528</a>
<a class='no' id='n529' name='n529' href='#n529'>529</a>
<a class='no' id='n530' name='n530' href='#n530'>530</a>
<a class='no' id='n531' name='n531' href='#n531'>531</a>
<a class='no' id='n532' name='n532' href='#n532'>532</a>
<a class='no' id='n533' name='n533' href='#n533'>533</a>
<a class='no' id='n534' name='n534' href='#n534'>534</a>
<a class='no' id='n535' name='n535' href='#n535'>535</a>
<a class='no' id='n536' name='n536' href='#n536'>536</a>
<a class='no' id='n537' name='n537' href='#n537'>537</a>
<a class='no' id='n538' name='n538' href='#n538'>538</a>
<a class='no' id='n539' name='n539' href='#n539'>539</a>
<a class='no' id='n540' name='n540' href='#n540'>540</a>
<a class='no' id='n541' name='n541' href='#n541'>541</a>
<a class='no' id='n542' name='n542' href='#n542'>542</a>
<a class='no' id='n543' name='n543' href='#n543'>543</a>
<a class='no' id='n544' name='n544' href='#n544'>544</a>
<a class='no' id='n545' name='n545' href='#n545'>545</a>
<a class='no' id='n546' name='n546' href='#n546'>546</a>
<a class='no' id='n547' name='n547' href='#n547'>547</a>
<a class='no' id='n548' name='n548' href='#n548'>548</a>
<a class='no' id='n549' name='n549' href='#n549'>549</a>
<a class='no' id='n550' name='n550' href='#n550'>550</a>
<a class='no' id='n551' name='n551' href='#n551'>551</a>
<a class='no' id='n552' name='n552' href='#n552'>552</a>
<a class='no' id='n553' name='n553' href='#n553'>553</a>
<a class='no' id='n554' name='n554' href='#n554'>554</a>
</pre></td>
<td class='lines'><pre><code>From 64f9449656dbbb718d53a54ed8e7904e289280ec Mon Sep 17 00:00:00 2001
From: Bastien Nocera &lt;hadess@hadess.net&gt;
Date: Fri, 30 Dec 2011 12:34:29 +0100
Subject: [PATCH] Add sixaxis cable-pairing plugin

Implement the old "sixpair" using libudev and libusb-1.0.

When a Sixaxis device is plugged in, events are filtered, and
the device is selected, poked around to set the default Bluetooth
address, and added to the database of the current default adapter.
---
 Makefile.am     |    9 +-
 acinclude.m4    |   16 +++
 configure.ac    |    1 +
 plugins/cable.c |  382 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/adapter.c   |   19 +++
 src/adapter.h   |    3 +
 6 files changed, 428 insertions(+), 2 deletions(-)
 create mode 100644 plugins/cable.c

diff --git a/Makefile.am b/Makefile.am
index 53fcbe9..f831a72 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -238,6 +238,11 @@ builtin_sources += thermometer/main.c \
 endif
 
 
+if CABLE
+builtin_modules += cable
+builtin_sources += plugins/cable.c
+endif
+
 builtin_modules += hciops mgmtops
 builtin_sources += plugins/hciops.c plugins/mgmtops.c
 
@@ -306,7 +311,7 @@ src_bluetoothd_SOURCES = $(gdbus_sources) $(builtin_sources) \
 			src/event.h src/event.c \
 			src/oob.h src/oob.c src/eir.h src/eir.c
 src_bluetoothd_LDADD = lib/libbluetooth-private.la @GLIB_LIBS@ @DBUS_LIBS@ \
-								-ldl -lrt
+								@CABLE_LIBS@ -ldl -lrt
 src_bluetoothd_LDFLAGS = $(AM_LDFLAGS) -Wl,--export-dynamic \
 				-Wl,--version-script=$(srcdir)/src/bluetooth.ver
 
@@ -428,7 +433,7 @@ EXTRA_DIST += doc/manager-api.txt \
 
 AM_YFLAGS = -d
 
-AM_CFLAGS += @DBUS_CFLAGS@ @GLIB_CFLAGS@
+AM_CFLAGS += @DBUS_CFLAGS@ @GLIB_CFLAGS@ @CABLE_CFLAGS@
 
 INCLUDES = -I$(builddir)/lib -I$(builddir)/src -I$(srcdir)/src \
 			-I$(srcdir)/audio -I$(srcdir)/sbc -I$(srcdir)/gdbus \
diff --git a/acinclude.m4 b/acinclude.m4
index 6505ad3..3f59989 100644
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -139,6 +139,12 @@ AC_DEFUN([AC_PATH_UDEV], [
 	AC_SUBST(UDEV_LIBS)
 ])
 
+AC_DEFUN([AC_PATH_CABLE], [
+	PKG_CHECK_MODULES(CABLE, libudev libusb-1.0, cable_found=yes, cable_found=no)
+	AC_SUBST(CABLE_CFLAGS)
+	AC_SUBST(CABLE_LIBS)
+])
+
 AC_DEFUN([AC_PATH_SNDFILE], [
 	PKG_CHECK_MODULES(SNDFILE, sndfile, sndfile_found=yes, sndfile_found=no)
 	AC_SUBST(SNDFILE_CFLAGS)
@@ -176,6 +182,7 @@ AC_DEFUN([AC_ARG_BLUEZ], [
 	sndfile_enable=${sndfile_found}
 	hal_enable=no
 	usb_enable=${usb_found}
+	cable_enable=${cable_found}
 	alsa_enable=${alsa_found}
 	gstreamer_enable=${gstreamer_found}
 	audio_enable=yes
@@ -265,6 +272,10 @@ AC_DEFUN([AC_ARG_BLUEZ], [
 		usb_enable=${enableval}
 	])
 
+	AC_ARG_ENABLE(cable, AC_HELP_STRING([--enable-cable], [enable DeviceKit support]), [
+		cable_enable=${enableval}
+	])
+
 	AC_ARG_ENABLE(tools, AC_HELP_STRING([--enable-tools], [install Bluetooth utilities]), [
 		tools_enable=${enableval}
 	])
@@ -366,6 +377,10 @@ AC_DEFUN([AC_ARG_BLUEZ], [
 		AC_DEFINE(HAVE_LIBUSB, 1, [Define to 1 if you have USB library.])
 	fi
 
+	if (test "${cable_enable}" = "yes" &amp;&amp; test "${cable_found}" = "yes"); then
+		AC_DEFINE(HAVE_CABLE, 1, [Define to 1 if you have libcable.])
+	fi
+
 	AM_CONDITIONAL(SNDFILE, test "${sndfile_enable}" = "yes" &amp;&amp; test "${sndfile_found}" = "yes")
 	AM_CONDITIONAL(USB, test "${usb_enable}" = "yes" &amp;&amp; test "${usb_found}" = "yes")
 	AM_CONDITIONAL(SBC, test "${alsa_enable}" = "yes" || test "${gstreamer_enable}" = "yes" ||
@@ -398,4 +413,5 @@ AC_DEFUN([AC_ARG_BLUEZ], [
 	AM_CONDITIONAL(DBUSOOBPLUGIN, test "${dbusoob_enable}" = "yes")
 	AM_CONDITIONAL(WIIMOTEPLUGIN, test "${wiimote_enable}" = "yes")
 	AM_CONDITIONAL(GATTMODULES, test "${gatt_enable}" = "yes")
+	AM_CONDITIONAL(CABLE, test "${cable_enable}" = "yes" &amp;&amp; test "${cable_found}" = "yes")
 ])
diff --git a/configure.ac b/configure.ac
index 48b181e..45a4b15 100644
--- a/configure.ac
+++ b/configure.ac
@@ -46,6 +46,7 @@ AC_PATH_GSTREAMER
 AC_PATH_USB
 AC_PATH_UDEV
 AC_PATH_SNDFILE
+AC_PATH_CABLE
 AC_PATH_OUI
 AC_PATH_READLINE
 AC_PATH_CHECK
diff --git a/plugins/cable.c b/plugins/cable.c
new file mode 100644
index 0000000..fe758db
--- /dev/null
+++ b/plugins/cable.c
@@ -0,0 +1,382 @@
+/*
+ *
+ *  BlueZ - Bluetooth protocol stack for Linux
+ *
+ *  Copyright (C) 2009  Bastien Nocera &lt;hadess@hadess.net&gt;
+ *
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include &lt;config.h&gt;
+#endif
+
+#include &lt;glib.h&gt;
+#define LIBUDEV_I_KNOW_THE_API_IS_SUBJECT_TO_CHANGE 1
+#include &lt;libudev.h&gt;
+#include &lt;dbus/dbus.h&gt;
+#include &lt;bluetooth/bluetooth.h&gt;
+#include &lt;bluetooth/sdp.h&gt;
+#include &lt;libusb.h&gt;
+
+#include "plugin.h"
+#include "log.h"
+
+#include "adapter.h"
+#include "manager.h"
+#include "device.h"
+
+#include "storage.h"
+#include "sdp_lib.h"
+
+/* Vendor and product ID for the Sixaxis PS3 controller */
+#define VENDOR 0x054c
+#define PRODUCT 0x0268
+#define SIXAXIS_PNP_RECORD "3601920900000A000100000900013503191124090004350D35061901000900113503190011090006350909656E09006A0901000900093508350619112409010009000D350F350D350619010009001335031900110901002513576972656C65737320436F6E74726F6C6C65720901012513576972656C65737320436F6E74726F6C6C6572090102251B536F6E7920436F6D707574657220456E7465727461696E6D656E740902000901000902010901000902020800090203082109020428010902052801090206359A35980822259405010904A101A102850175089501150026FF00810375019513150025013500450105091901291381027501950D0600FF8103150026FF0005010901A10075089504350046FF0009300931093209358102C0050175089527090181027508953009019102750895300901B102C0A1028502750895300901B102C0A10285EE750895300901B102C0A10285EF750895300901B102C0C0090207350835060904090901000902082800090209280109020A280109020B09010009020C093E8009020D280009020E2800"
+#define HID_UUID "00001124-0000-1000-8000-00805f9b34fb"
+
+static struct btd_device *create_cable_association(DBusConnection *conn,
+						    struct btd_adapter *adapter,
+						    const char *name,
+						    const char *address,
+						    guint32 vendor_id,
+						    guint32 product_id,
+						    const char *pnp_record)
+{
+	sdp_record_t *rec;
+	struct btd_device *device;
+	bdaddr_t src, dst;
+	char srcaddr[18];
+
+	device = adapter_find_device(adapter, address);
+	if (device == NULL) {
+		device = device_create(conn, adapter, address, BDADDR_BREDR);
+		if (device != NULL)
+			adapter_create_device_for_device(conn, adapter, device);
+	}
+	if (device != NULL) {
+		device_set_temporary(device, FALSE);
+		device_set_name(device, name);
+	}
+
+	str2ba(address, &amp;dst);
+	adapter_get_address(adapter, &amp;src);
+	ba2str(&amp;src, srcaddr);
+
+	write_device_name(&amp;dst, &amp;src, (char *) name);
+
+	/* Store the device's SDP record */
+	rec = record_from_string(pnp_record);
+	store_record(srcaddr, address, rec);
+	sdp_record_free(rec);
+	/* Set the device id */
+	store_device_id(srcaddr, address, 0xffff, vendor_id, product_id, 0);
+	/* Don't write a profile, it will be updated when the device connects */
+
+	write_trust(srcaddr, address, "[all]", TRUE);
+
+	return device;
+}
+
+static char *get_bdaddr(libusb_device_handle *devh, int itfnum)
+{
+	unsigned char msg[17];
+	char *address;
+	int res;
+
+	res = libusb_control_transfer(devh,
+				      LIBUSB_ENDPOINT_IN | LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE,
+				      0x01, 0x03f2, itfnum,
+				      (void*) msg, sizeof(msg),
+				      5000);
+
+	if (res &lt; 0) {
+		DBG("Getting the device Bluetooth address failed");
+		return NULL;
+	}
+
+	address = g_strdup_printf("%02X:%02X:%02X:%02X:%02X:%02X",
+				  msg[4], msg[5], msg[6], msg[7], msg[8], msg[9]);
+
+	DBG("Device Bluetooth address: %s\n", address);
+
+	return address;
+}
+
+static gboolean set_master_bdaddr(libusb_device_handle *devh, int itfnum, char *host)
+{
+	unsigned char msg[8];
+	int mac[6];
+	int res;
+
+	if (sscanf(host, "%X:%X:%X:%X:%X:%X",
+		   &amp;mac[0],&amp;mac[1],&amp;mac[2],&amp;mac[3],&amp;mac[4],&amp;mac[5]) != 6) {
+		return FALSE;
+	}
+
+	msg[0] = 0x01;
+	msg[1] = 0x00;
+	msg[2] = mac[0];
+	msg[3] = mac[1];
+	msg[4] = mac[2];
+	msg[5] = mac[3];
+	msg[6] = mac[4];
+	msg[7] = mac[5];
+
+	res = libusb_control_transfer(devh,
+				      LIBUSB_ENDPOINT_OUT | LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE,
+				      0x09, 0x03f5, itfnum,
+				      (void*) msg, sizeof(msg),
+				      5000);
+
+	if (res &lt; 0) {
+		DBG("Setting the master Bluetooth address failed");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+static void handle_usb_device(struct btd_adapter *adapter,
+			      libusb_device *dev,
+			      struct libusb_config_descriptor *cfg,
+			      int itfnum,
+			      const struct libusb_interface_descriptor *alt)
+{
+	DBusConnection *conn;
+	libusb_device_handle *devh;
+	char *device_bdaddr;
+	char adapter_bdaddr[18];
+	struct btd_device *device;
+	bdaddr_t dst;
+
+	device_bdaddr = NULL;
+	conn = dbus_bus_get(DBUS_BUS_SYSTEM, NULL);
+	if (conn == NULL) {
+		DBG("Failed to get on the bus");
+		return;
+	}
+
+	if (libusb_open(dev, &amp;devh) &lt; 0) {
+		DBG("Can't open device");
+		goto bail;
+	}
+	libusb_detach_kernel_driver(devh, itfnum);
+
+	if (libusb_claim_interface(devh, itfnum) &lt; 0) {
+		DBG("Can't claim interface %d", itfnum);
+		goto bail;
+	}
+
+	device_bdaddr = get_bdaddr(devh, itfnum);
+	if (device_bdaddr == NULL) {
+		DBG("Failed to get the Bluetooth address from the device");
+		goto bail;
+	}
+
+	device = create_cable_association(conn,
+					  adapter,
+					  "PLAYSTATION(R)3 Controller",
+					  device_bdaddr,
+					  VENDOR, PRODUCT, SIXAXIS_PNP_RECORD);
+	btd_device_add_uuid(device, HID_UUID);
+
+	adapter_get_address(adapter, &amp;dst);
+	ba2str(&amp;dst, adapter_bdaddr);
+	DBG("Adapter bdaddr %s", adapter_bdaddr);
+
+	if (set_master_bdaddr(devh, itfnum, adapter_bdaddr) == FALSE) {
+		DBG("Failed to set the master Bluetooth address");
+		goto bail;
+	}
+
+bail:
+	dbus_connection_unref(conn);
+	g_free(device_bdaddr);
+	libusb_release_interface(devh, itfnum);
+	/* We ignore errors from the reattach, as there's nothing we
+	 * can do about it */
+	libusb_attach_kernel_driver(devh, itfnum);
+	if (devh != NULL)
+		libusb_close(devh);
+}
+
+static void handle_device_plug(struct udev_device *udevice)
+{
+	struct btd_adapter *adapter;
+	guint i;
+
+	libusb_device **list, *usbdev;
+	ssize_t num_devices;
+	struct libusb_device_descriptor desc;
+	guint8 j;
+
+	if (g_strcmp0(udev_device_get_property_value(udevice, "ID_SERIAL"),
+		      "Sony_PLAYSTATION_R_3_Controller") != 0)
+		return;
+	/* Don't look at events with an associated driver */
+	if (udev_device_get_property_value(udevice, "ID_USB_DRIVER") != NULL)
+		return;
+
+	DBG("Found Sixaxis device");
+
+	/* Look for the default adapter */
+	adapter = manager_get_default_adapter();
+	if (adapter == NULL)
+		return;
+
+	/* Look for the USB device */
+	libusb_init(NULL);
+
+	num_devices = libusb_get_device_list(NULL, &amp;list);
+	if (num_devices &lt; 0) {
+		DBG("libusb_get_device_list failed");
+		return;
+	}
+
+	usbdev = NULL;
+	for (i = 0; i &lt; num_devices; i++) {
+		char *path;
+
+		path = g_strdup_printf("%s/%03d/%03d", "/dev/bus/usb",
+				       libusb_get_bus_number(list[i]),
+				       libusb_get_device_address(list[i]));
+		if (g_strcmp0(path, udev_device_get_devnode(udevice)) == 0) {
+			g_free(path);
+			usbdev = libusb_ref_device(list[i]);
+			break;
+		}
+		g_free(path);
+	}
+
+	libusb_free_device_list(list, TRUE);
+	if (usbdev == NULL) {
+		DBG("Found a Sixaxis, but couldn't find it via libusb");
+		goto out;
+	}
+
+	if (libusb_get_device_descriptor(usbdev, &amp;desc) &lt; 0) {
+		DBG("libusb_get_device_descriptor() failed");
+		goto out;
+	}
+
+	/* Look for the interface number that interests us */
+	for (j = 0; j &lt; desc.bNumConfigurations; j++) {
+		struct libusb_config_descriptor *config;
+		guint8 k;
+
+		if (libusb_get_config_descriptor(usbdev, j, &amp;config) &lt; 0) {
+			DBG("Failed to get config descriptor %d", j);
+			continue;
+		}
+
+		for (k = 0; k &lt; config-&gt;bNumInterfaces; k++) {
+			const struct libusb_interface *itf = &amp;config-&gt;interface[k];
+			int l;
+
+			for (l = 0; l &lt; itf-&gt;num_altsetting ; l++) {
+				struct libusb_interface_descriptor alt;
+
+				alt = itf-&gt;altsetting[l];
+				if (alt.bInterfaceClass == 3) {
+					handle_usb_device(adapter, usbdev, config, l, &amp;alt);
+				}
+			}
+		}
+	}
+
+out:
+	if (usbdev != NULL)
+		libusb_unref_device(usbdev);
+	libusb_exit(NULL);
+}
+
+static gboolean device_event_idle(struct udev_device *udevice)
+{
+	handle_device_plug(udevice);
+	udev_device_unref(udevice);
+	return FALSE;
+}
+
+static struct udev *ctx = NULL;
+static struct udev_monitor *monitor = NULL;
+static guint watch_id = 0;
+
+static gboolean
+monitor_event(GIOChannel *source,
+	      GIOCondition condition,
+	      gpointer data)
+{
+	struct udev_device *udevice;
+
+	udevice = udev_monitor_receive_device(monitor);
+	if (udevice == NULL)
+		goto out;
+	if (g_strcmp0(udev_device_get_action(udevice), "add") != 0)
+		goto out;
+
+	g_timeout_add_seconds(1, (GSourceFunc) device_event_idle, udevice);
+
+out:
+	return TRUE;
+}
+
+
+static int cable_init(void)
+{
+	GIOChannel *channel;
+
+	DBG("Setup cable plugin");
+
+	ctx = udev_new();
+	monitor = udev_monitor_new_from_netlink(ctx, "udev");
+	if (monitor == NULL) {
+		error ("Could not get udev monitor");
+		return -1;
+	}
+
+	/* Listen for newly connected usb device */
+	udev_monitor_filter_add_match_subsystem_devtype(monitor,
+							"usb", NULL);
+	udev_monitor_enable_receiving(monitor);
+
+	channel = g_io_channel_unix_new(udev_monitor_get_fd(monitor));
+	watch_id = g_io_add_watch(channel, G_IO_IN, monitor_event, NULL);
+	g_io_channel_unref(channel);
+
+	return 0;
+}
+
+static void cable_exit(void)
+{
+	DBG("Cleanup cable plugin");
+
+	if (watch_id != 0) {
+		g_source_remove(watch_id);
+		watch_id = 0;
+	}
+	if (monitor != NULL) {
+		udev_monitor_unref(monitor);
+		monitor = NULL;
+	}
+	if (ctx != NULL) {
+		udev_unref(ctx);
+		ctx = NULL;
+	}
+}
+
+BLUETOOTH_PLUGIN_DEFINE(cable, VERSION,
+			BLUETOOTH_PLUGIN_PRIORITY_DEFAULT, cable_init, cable_exit)
diff --git a/src/adapter.c b/src/adapter.c
index 6e04faf..0488891 100644
--- a/src/adapter.c
+++ b/src/adapter.c
@@ -952,6 +952,25 @@ static struct btd_device *adapter_create_device(DBusConnection *conn,
 	return device;
 }
 
+void adapter_create_device_for_device(DBusConnection *conn,
+				      struct btd_adapter *adapter,
+				      struct btd_device *device)
+{
+	const char *path;
+
+	device_set_temporary(device, TRUE);
+
+	adapter-&gt;devices = g_slist_append(adapter-&gt;devices, device);
+
+	path = device_get_path(device);
+	g_dbus_emit_signal(conn, adapter-&gt;path,
+			ADAPTER_INTERFACE, "DeviceCreated",
+			DBUS_TYPE_OBJECT_PATH, &amp;path,
+			DBUS_TYPE_INVALID);
+
+	adapter_update_devices(adapter);
+}
+
 void adapter_remove_device(DBusConnection *conn, struct btd_adapter *adapter,
 						struct btd_device *device,
 						gboolean remove_storage)
diff --git a/src/adapter.h b/src/adapter.h
index b7ea62b..ac0aa2e 100644
--- a/src/adapter.h
+++ b/src/adapter.h
@@ -114,6 +114,9 @@ void adapter_update_found_devices(struct btd_adapter *adapter,
 					uint8_t *data, uint8_t data_len);
 void adapter_emit_device_found(struct btd_adapter *adapter,
 						struct remote_dev_info *dev);
+void adapter_create_device_for_device(DBusConnection *conn,
+				      struct btd_adapter *adapter,
+				      struct btd_device *device);
 void adapter_mode_changed(struct btd_adapter *adapter, uint8_t scan_mode);
 int adapter_set_name(struct btd_adapter *adapter, const char *name);
 void adapter_name_changed(struct btd_adapter *adapter, const char *name);
-- 
1.7.10.2

</code></pre></td></tr></table>
</div> <!-- class=content -->
<div class='footer'>generated  by cgit v0.9.0.3 at 2012-11-07 00:55:29 (GMT)</div>
</div> <!-- id=cgit -->
</body>
</html>
